### DISCUSSION TITLE
Spi Matrix with Pi4j V2.0

### DISCUSSION THREAD

**Discussion by matti318** [Pi4J/pi4j#167] [Q&A] [answered]
Hello, I would like to operate a matrix 8x8 on my Raspberry with Pi4j V2.0, unfortunately there are instructions only with crowpi. But I have Raspbian on the Raspberry. Is there a tutorial on how to do it easily with Raspbian and the Matrix?
greeting

**Comment by FDelporte** [comment]
CrowPi OS is actually the same as Raspbian (now known as Raspberry Pi OS) with some additional settings. So you should be able to run the exact same code on both OS versions.

One thing to take into account: if using SPI, you may need to set your Pi4J context yourself instead to make sure the right libraries are used to communicate with SPI over PiGpio as is done on https://github.com/Pi4J/pi4j-example-crowpi/blob/62977aa4564bef6edd4fc50fa6c5feeafaadc0b0/src/main/java/com/pi4j/crowpi/helpers/CrowPiPlatform.java#L29

**Comment by matti318** [comment]
Here is my code. It should be a simple code to test. It worked fine with version 1.4.
`
import com.pi4j.Pi4J;
import com.pi4j.io.spi.Spi;
import com.pi4j.io.spi.SpiConfig;
import java.io.IOException;
import com.pi4j.context.Context;

public class Matrix {


    public static void main(String[] args) throws InterruptedException, IOException{
        var pi4j = Pi4J.newAutoContext();
        Max7219 neu = new Max7219(pi4j.create(buildSpiConfig(pi4j, 0, 8000000)));
        System.out.println("hello");

        neu.initial();
        neu.aus();
        
        neu.execute((byte)0, (byte)129);
//        neu.execute((byte)0x0c, (byte)0x00);
        
        
    }
    private static SpiConfig buildSpiConfig(Context pi4j, int channel, int baud) {
        return Spi.newConfigBuilder(pi4j)
            .id("SPI" + channel)
            .name("LED Matrix")
            .address(channel)
 //           .baud(baud)
            .build();}
}

class Max7219{
    protected final Spi spi;
    public Max7219(Spi spi) {
        this.spi = spi;}
    
    
     void initial() throws IOException, InterruptedException{ // hier wird alles initialisiert
        execute((byte) 0x09, (byte) 0x00); // decode mode
        execute((byte)0x0a, (byte)15); // intensität 1-15
        execute((byte)0x0b, (byte)0x07); // scan limit, 07 ist stehet fuer 8 Reihen
        execute((byte)0x0c, (byte)0x01); // steht fuer shutdown, 01 ist normal mod
        execute((byte)0x0f, (byte)0x00); //display test, 00 ist aus
    }
    
    
     void aus()  throws IOException, InterruptedException{  //schaltet erst einmal alles aus
        int end = 8;
        for (int i = 1; i <=end; i++) {
            spi.write((byte)i, (byte)0);
        System.out.println(i);}
        }
     public void execute(byte command, byte data) {
            spi.write(command, data);
    
    
    
    
}
}
`

What else do you have to change?

**Reply by FDelporte** [reply]
can you replace the newAutoContext() with PiGpio provider as in https://github.com/Pi4J/pi4j-example-crowpi/blob/62977aa4564bef6edd4fc50fa6c5feeafaadc0b0/src/main/java/com/pi4j/crowpi/helpers/CrowPiPlatform.java#L29?


```
Pi4J.newContextBuilder()
            .noAutoDetect()
            .add(
                PiGpioDigitalInputProvider.newInstance(piGpio),
                PiGpioDigitalOutputProvider.newInstance(piGpio),
                PiGpioPwmProvider.newInstance(piGpio),
                PiGpioI2CProvider.newInstance(piGpio),
                PiGpioSerialProvider.newInstance(piGpio),
                PiGpioSpiProvider.newInstance(piGpio)
            )
            .build();
```

**Comment by matti318** [comment]
How do you mean that exactly? I have to pass parameters for the Max7219 class for the constructor.


private static SpiConfig buildSpiConfig(Context pi4j, int channel, int baud) {
		        final var piGpio = PiGpio.newNativeInstance();
		        return Pi4J.newContextBuilder()
	                .noAutoDetect()
	                .add(new Matrix())
	                .add(
	                PiGpioDigitalInputProvider.newInstance(piGpio),
	                PiGpioDigitalOutputProvider.newInstance(piGpio),
	                PiGpioPwmProvider.newInstance(piGpio),
	                PiGpioI2CProvider.newInstance(piGpio),
	                PiGpioSerialProvider.newInstance(piGpio),
	                PiGpioSpiProvider.newInstance(piGpio)
	            )
	               .build();
	    }

Does not work like that.

**Comment by matti318** [comment, marked as answer]
Now the code works. Is the code okay like that?

```
import com.pi4j.Pi4J;
import com.pi4j.io.spi.Spi;
import com.pi4j.io.spi.SpiConfig;
import java.io.IOException;
import com.pi4j.context.Context;
import com.pi4j.library.pigpio.PiGpio;
import com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalInputProvider;
import com.pi4j.plugin.pigpio.provider.gpio.digital.PiGpioDigitalOutputProvider;
import com.pi4j.plugin.pigpio.provider.i2c.PiGpioI2CProvider;
import com.pi4j.plugin.pigpio.provider.pwm.PiGpioPwmProvider;
import com.pi4j.plugin.pigpio.provider.serial.PiGpioSerialProvider;
import com.pi4j.plugin.pigpio.provider.spi.PiGpioSpiProvider;


public class Matrix {


	public static void main(String[] args) throws InterruptedException, IOException{
		final var piGpio = PiGpio.newNativeInstance();
		var pi4j = Pi4J.newContextBuilder()
	            .noAutoDetect()
	            .add(
	                PiGpioDigitalInputProvider.newInstance(piGpio),
	                PiGpioDigitalOutputProvider.newInstance(piGpio),
	                PiGpioPwmProvider.newInstance(piGpio),
	                PiGpioI2CProvider.newInstance(piGpio),
	                PiGpioSerialProvider.newInstance(piGpio),
	                PiGpioSpiProvider.newInstance(piGpio)
	            )
	            .build();
		
		Max7219 neu = new Max7219(pi4j.create(buildSpiConfig(pi4j, 0, 8000000)));//(buildSpiConfig(pi4j, 0, 8000000)));
		System.out.println("hhello");
		neu.initial();
		neu.aus();
		neu.execute((byte)1, (byte)129);	
	}
	
	private static SpiConfig  buildSpiConfig(Context pi4j, int channel, int baud) {
           return Spi.newConfigBuilder(pi4j)
              .id("SPI" + channel)
              .name("Matrix")
              .address(channel)
              .baud(baud)
              .build();
       }
}

class Max7219{
	protected final Spi spi;
	public Max7219(Spi spi) {
		this.spi = spi;
        }
	
	void initial() throws IOException, InterruptedException{ // hier wird alles initialisiert
		execute((byte) 0x09, (byte) 0x00); // decode mode
		execute((byte)0x0a, (byte)1); // intensität 1-15
		execute((byte)0x0b, (byte)0x07); // scan limit, 07 ist stehet fuer 8 Reihen
		execute((byte)0x0c, (byte)0x01); // steht fuer shutdown, 01 ist normal mod
		execute((byte)0x0f, (byte)0x00); //display test, 00 ist aus
	}
	
	void aus()  throws IOException, InterruptedException{  //schaltet erst einmal alles aus
		int end = 8;
		for (int i = 1; i <=end; i++) {
			execute((byte)i, (byte)0);
			System.out.println("nuna" + i);
                }
        }

	 public void execute(byte command, byte data){
	     spi.write(command, data);      
        }
}
```

**Reply by FDelporte** [reply]
Hey, indeed looks OK, without running your code myself ;-)
Can you maybe try to only include the SpiProvider, as I guess that's the only type you are using?

```
var pi4j = Pi4J.newContextBuilder()
            .noAutoDetect()
            .add(
                PiGpioSpiProvider.newInstance(piGpio)
            )
            .build();
```

**Comment by matti318** [comment]
OK, that works. Thanks very much.

**Comment by matti318** [comment]
OK, that works. Thanks very much.

**Comment by FDelporte** [comment]
Thanks for your feedback!
I added this to the documentation page on https://pi4j.com/documentation/io-examples/spi/

**Comment by matti318** [comment]
thanks
